channels/security/websocket.py

```diff
@@ -21,27 +24,134 @@ def __call__(self, scope):
+        """
+        Validate the given origin for this site.
+
+        Check than the origin looks valid and matches the origin pattern in
+        specified list `` allowed_origins``. Any pattern begins with a scheme.
+        After the scheme there must be a domain. Any domain beginning with a period
+        corresponds to the domain and all its subdomains (for example, ``http://.example.com``
+        ``http://example.com`` and any subdomain).After the domain there must be a port,
+        but it can be omitted. `` * `` matches anything and anything
+        else must match exactly.
+
+        Note. This function assumes that the given origin has a schema, domain and port,
+        but port is optional.
+
+        Returns `` True`` for a valid host, `` False`` otherwise.
+        """
+        return any(
+            pattern == "*" or self.is_allowed_origin(parsed_origin, pattern)
+            for pattern in self.allowed_origins
+        )
+
+    def is_allowed_origin(self, parsed_origin, pattern):
+        """
+        Returns ``True`` if the origin is either an exact match or a match
+        to the wildcard pattern. Compares scheme, domain, port of origin and pattern.
+
+        Any pattern can be begins with a scheme. After the scheme must be a domain,
+        or just domain without scheme.
+        Any domain beginning with a period corresponds to the domain and all
+        its subdomains (for example, ``.example.com`` ``example.com``
+        and any subdomain). Also with scheme (for example, ``http://.example.com``
+        ``http://exapmple.com``). After the domain there must be a port,
+        but it can be omitted.
+
+        Also, if scheme is ``https``, verifies security certificate.
+
+        Note. This function assumes that the given origin has a schema, domain and port,
+        or only domain. Port is optional.
+
+        Raises SSLError if origin used an invalid security certificate.
+
+        Raises ConnectionRefusedError if the connection is rejected.
+        Occurs when connected to a port that is not listening.
+
+        Raises socket.gaierror if name or service not known.
+        """
+        # Get ResultParse object
+        parsed_pattern = urlparse(pattern.lower(), scheme=None)
+        if parsed_pattern.scheme is None:
+            pattern_hostname = urlparse("//" + pattern).hostname or pattern
+            return is_same_domain(parsed_origin.hostname, pattern_hostname)
+        # Get origin.port or default ports for origin or None
+        origin_port = self.get_origin_port(parsed_origin)
+        # Get pattern.port or default ports for pattern or None
+        pattern_port = self.get_origin_port(parsed_pattern)
+        # Compares hostname, scheme, ports of pattern and origin
+        if parsed_pattern.scheme == parsed_origin.scheme and origin_port == pattern_port \
+                and is_same_domain(parsed_origin.hostname, parsed_pattern.hostname):
+            # Check ssl security certificate
```

Comment:
I do not think we should try to check the certificate in this code at all, and definitely not by default.